### ISSUE TITLE
Can't depend on "."

### ISSUE THREAD

**Issue by colincross** [ninja-build/ninja#1075]
When handling globs, Blueprint emits dependencies to regenerate build.ninja if a directory is modified, in case the modification was to add a missing path.  For example, when globbing for dir/*, it will add a dependency on "dir" if it exists (to handle the case where a file is added or delete from dir), or "." if a does not exist (to handle the case where dir is added).  The first case works fine, but the second case fails with:
`ninja: error: rebuilding 'build.ninja': path canonicalizes to the empty path`

"." seems like a perfectly reasonable path, and should canonicalize to ".".

**Comment by martine** [comment]
I can't quite remember why we have this check, but I think it's to prevent accidentally incorrect paths, such as when generators accidentally produce the empty string.  E.g.

```
outfile = foobar.so
build $outfilex: ...
# oops, typoed the variable name
```

The other accident is paths like `foo/..` which I don't think people intend to provide.

This is all a long way of asking: would it still work for you if we explicitly allowed the string "." as a path but disallowed the other cases?

**Comment by colincross** [comment]
Allowing "." would be fine for my issue, although I think disallowing `foo/..` is unnecessary - depending on `foo/..` is clearly depending on a directory, not a file, and is not much different than `foo/bar/..`.

The empty string case is already caught by the first check check for *len == 0 in CanonicalizePath.

**Comment by colincross** [comment]
The "canonicalizes to the empty path" error comes from e9087f5cdf2c8485b5b0a6858782bb02dbffbcdd which was preventing *len being set to -1, and not actually trying to prevent ".".

**Comment by nico** [comment]
fixed by #1076
